Specific Error Message or Note for Final Expression being Loop and Return not Unit

```rust
fn loop_ending() -> i32 {
    loop {
        if false { break; }
        return 42;
    }
}
```

produces

```
error[E0308]: mismatched types
 --> <anon>:2:5
  |
2 |       loop {
  |  _____^ starting here...
3 | |         if false { break; }
4 | |         return 42;
5 | |     }
  | |_____^ ...ending here: expected i32, found ()
  |
  = note: expected type `i32`
  = note:    found type `()`

error: aborting due to previous error
```

But the compiler can check that the final expression is a loop and must always evaluate into `()`, so we can recommend more advanced information.

This has caused [real confusion](https://www.reddit.com/r/rust/comments/5v163r/loop_has_to_return_a_stdrestultresult_since_when/).

Another example (Derived from #18390):

```rust
fn main() {
    let _ = foo();
}

fn foo() -> i32 {
    let x = 5;
    match x {
        1 => {
            1
        },
        2 => {
            2
        },
        3 => {
            3
        },
        4 => {
            4
        },
        5 => {
            5
        },
        _ => {
            0
        },
    };
    match x {
        1 => {
            return 1
        },
        2 => {
            return 2
        },
        3 => {
            return 3
        },
        4 => {
            return 4
        },
        5 => {
            return 5
        },
        _ => { }
    }
}
```

gives:

```
error[E0308]: mismatched types
  --> src/main.rs:5:17
   |
5  |   fn foo() -> i32 {
   |  _________________^
6  | |     let x = 5;
7  | |     match x {
8  | |         1 => {
...  |
44 | |     }
45 | | }
   | |_^ expected i32, found ()
   |
   = note: expected type `i32`
              found type `()`
```

I guess the key point in both cases is that the user is writing more "C-style" code and not *expecting* to make use of the "implicit return" here, so the error is sort of surprising.

It seems to me that we could cover this by a similar rule to the one we use to suggest removing a semicolon -- in particular, if the *expected type* is not unit, but we are returning unit from a tail expression, and that tail expression is some kind of "control-flow" structure, we might point to the end of it and say something like "no return along this path":

```rust
error[E0308]: no return along this path
 --> src/main.rs:3:20
  |
3 |         if false { break; }
  |                    ^^^^^ following this path, no value is ever returned
  |
```

This will require a bit more state to be plumbed around, I imagine, not entirely sure what (and I'm not entirely sure *where* the error should be reported -- on the break, as I showed above?).

The current output is even worse, IMO:

```
error[E0308]: mismatched types
 --> src/main.rs:3:20
  |
3 |         if false { break; }
  |                    ^^^^^ expected (), found i32
  |
  = note: expected type `()`
             found type `i32`
```

Current output:

```
error[E0308]: mismatched types
 --> src/lib.rs:3:20
  |
3 |         if false { break; }
  |                    ^^^^^
  |                    |
  |                    expected i32, found ()
  |                    help: give it a value of the expected type: `break 42`
  |
  = note: expected type `i32`
             found type `()`
```

--- a/src/librustc/traits/error_reporting.rs
+++ b/src/librustc/traits/error_reporting.rs
@@ -1169,7 +1169,7 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
     fn suggest_fn_call(
         &self,
         obligation: &PredicateObligation<'tcx>,
-        err: &mut DiagnosticBuilder<'tcx>,
+        err: &mut DiagnosticBuilder<'_>,
         trait_ref: &ty::Binder<ty::TraitRef<'tcx>>,
         points_at_arg: bool,
     ) {

--- a/src/librustc_typeck/check/demand.rs
+++ b/src/librustc_typeck/check/demand.rs
@@ -553,7 +553,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
 
     pub fn check_for_cast(
         &self,
-        err: &mut DiagnosticBuilder<'tcx>,
+        err: &mut DiagnosticBuilder<'_>,
         expr: &hir::Expr,
         checked_ty: Ty<'tcx>,
         expected_ty: Ty<'tcx>,

--- a/src/librustc_typeck/check/expr.rs
+++ b/src/librustc_typeck/check/expr.rs
@@ -583,7 +583,15 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 } else {
                     assert!(e_ty.is_unit());
                     let ty = coerce.expected_ty();
-                    coerce.coerce_forced_unit(self, &cause, &mut |err| {
+                    coerce.coerce_forced_unit(self, &cause, &mut |mut err| {
+                        self.suggest_mismatched_types_on_tail(
+                            &mut err,
+                            expr,
+                            ty,
+                            e_ty,
+                            cause.span,
+                            target_id,
+                        );
                         let val = match ty.kind {
                             ty::Bool => "true",
                             ty::Char => "'a'",

--- a/src/librustc_typeck/check/mod.rs
+++ b/src/librustc_typeck/check/mod.rs
@@ -4246,7 +4246,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
     /// - Possible missing return type if the return type is the default, and not `fn main()`.
     pub fn suggest_mismatched_types_on_tail(
         &self,
-        err: &mut DiagnosticBuilder<'tcx>,
+        err: &mut DiagnosticBuilder<'_>,
         expr: &'tcx hir::Expr,
         expected: Ty<'tcx>,
         found: Ty<'tcx>,
@@ -4273,7 +4273,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
     /// ```
     fn suggest_fn_call(
         &self,
-        err: &mut DiagnosticBuilder<'tcx>,
+        err: &mut DiagnosticBuilder<'_>,
         expr: &hir::Expr,
         expected: Ty<'tcx>,
         found: Ty<'tcx>,
@@ -4386,7 +4386,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
 
     pub fn suggest_ref_or_into(
         &self,
-        err: &mut DiagnosticBuilder<'tcx>,
+        err: &mut DiagnosticBuilder<'_>,
         expr: &hir::Expr,
         expected: Ty<'tcx>,
         found: Ty<'tcx>,
@@ -4454,7 +4454,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
     /// in the heap by calling `Box::new()`.
     fn suggest_boxing_when_appropriate(
         &self,
-        err: &mut DiagnosticBuilder<'tcx>,
+        err: &mut DiagnosticBuilder<'_>,
         expr: &hir::Expr,
         expected: Ty<'tcx>,
         found: Ty<'tcx>,
@@ -4498,7 +4498,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
     /// it suggests adding a semicolon.
     fn suggest_missing_semicolon(
         &self,
-        err: &mut DiagnosticBuilder<'tcx>,
+        err: &mut DiagnosticBuilder<'_>,
         expression: &'tcx hir::Expr,
         expected: Ty<'tcx>,
         cause_span: Span,
@@ -4537,7 +4537,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
     /// type.
     fn suggest_missing_return_type(
         &self,
-        err: &mut DiagnosticBuilder<'tcx>,
+        err: &mut DiagnosticBuilder<'_>,
         fn_decl: &hir::FnDecl,
         expected: Ty<'tcx>,
         found: Ty<'tcx>,
@@ -4603,7 +4603,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
     /// `.await` to the tail of the expression.
     fn suggest_missing_await(
         &self,
-        err: &mut DiagnosticBuilder<'tcx>,
+        err: &mut DiagnosticBuilder<'_>,
         expr: &hir::Expr,
         expected: Ty<'tcx>,
         found: Ty<'tcx>,

--- a/src/test/ui/type/type-error-break-tail.rs
+++ b/src/test/ui/type/type-error-break-tail.rs
@@ -1,0 +1,8 @@
+fn loop_ending() -> i32 {
+    loop {
+        if false { break; } //~ ERROR mismatched types
+        return 42;
+    }
+}
+
+fn main() {}

--- a/src/test/ui/type/type-error-break-tail.stderr
+++ b/src/test/ui/type/type-error-break-tail.stderr
@@ -1,0 +1,18 @@
+error[E0308]: mismatched types
+  --> $DIR/type-error-break-tail.rs:3:20
+   |
+LL | fn loop_ending() -> i32 {
+   |                     --- expected `i32` because of return type
+LL |     loop {
+LL |         if false { break; }
+   |                    ^^^^^
+   |                    |
+   |                    expected i32, found ()
+   |                    help: give it a value of the expected type: `break 42`
+   |
+   = note: expected type `i32`
+              found type `()`
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0308`.

